src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java

```diff
@@ -0,0 +1,218 @@
+                if (object.getClass().isArray()) {
+                    return new ArrayMap(object);
+                } else if (object instanceof Collection) {
+                    return new CollectionMap((Collection) object);
+                }
+            }
+            return super.coerce(object);
+        }
+
+    }
+
+    static class ArrayMap extends AbstractMap<Object, Object> implements Iterable<Object> {
+
+        private final Object array;
+
+        public ArrayMap(Object array) {
+            this.array = array;
+        }
+
+        @Override
+        public Object get(Object key) {
+            if (key instanceof Number) {
+                return Array.get(array, ((Number) key).intValue());
+            }
+            try {
+                int index = Integer.parseInt(key.toString());
+                return Array.get(array, index);
+            } catch (NumberFormatException nfe) {
+                // if it's not a number it is as if the key doesn't exist
+                return null;
+            }
+        }
+
+        @Override
+        public boolean containsKey(Object key) {
+            return get(key) != null;
+        }
+
+        @Override
+        public Set<Entry<Object, Object>> entrySet() {
+            int length = Array.getLength(array);
+            Map<Object, Object> map = new HashMap<>(length);
+            for (int i = 0; i < length; i++) {
+                map.put(i, Array.get(array, i));
+            }
+            return map.entrySet();
+        }
+
+        /**
+         * Returns an iterator over a set of elements of type T.
+         *
+         * @return an Iterator.
+         */
+        @Override
+        public Iterator<Object> iterator() {
+            return new Iter(array);
+        }
+
+        static class Iter implements Iterator<Object> {
+
+            private final Object array;
+            private final int length;
+            private int index;
+
+            public Iter(Object array) {
+                this.array = array;
+                this.length = Array.getLength(array);
+                this.index = 0;
+            }
+
+            @Override
+            public boolean hasNext() {
+                return index < length;
+            }
+
+            @Override
+            public Object next() {
+                return Array.get(array, index++);
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException("array iterator does not support removing elements");
+            }
+        }
+    }
+
+    static class CollectionMap extends AbstractMap<Object, Object> implements Iterable<Object> {
+
+        private final Collection col;
+
+        public CollectionMap(Collection col) {
+            this.col = col;
+        }
+
+        @Override
+        public Object get(Object key) {
+            if (key instanceof Number) {
+                return Iterables.get(col, ((Number) key).intValue());
```

Comment:
> I added generic collections here so we'll always fall back on a behaviour (even if for unsorted sets it's non-
> deterministic). that said, I can change this to only support arrays, lists and sorted sets...

For context: Despite having had trouble understanding the semantics here IMHO this is one of the cases where failing fast is best. Probably I've encountered too many instances where instead of reading the docs people just "did whatever worked on first try" only to run into issues later.